README.md

**This demo environment now runs only on Kubernetes**. 

To make use of this repo, you'll need to clone it locally - if you don't know how to do that, see the Github [Cloning a repository](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository) documentation. 

## Changes

To see what's changed in the current version, see [CHANGES.md](CHANGES.md)

## Running Locally

To run this locally, we recommend minikube. Additionally, this environment uses `skaffold` to orchestrate building the requisite containers and deploying into Kubernetes. 

_**NOTE**_ - with this version of the repository, we need to use a version of skaffold that's v1.24.0 or greater - the instructions below have been changed to reflect that.  

## Pre-Requisites

### On a Mac with homebrew ([Homebrew Installation Instructions](https://docs.brew.sh/Installation))

```
brew install minikube
brew install kubectl
brew install helm
curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-darwin-amd64 && \
chmod +x skaffold && sudo mv skaffold /usr/local/bin
```

### On a Mac with MacPorts ([MacPorts Installation Instructions](https://www.macports.org/install.php))
```
sudo port install minikube kubectl-1.20 helm-3.5
sudo port select --set helm helm3.5
sudo port select --set kubectl kubectl1.20
curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-darwin-amd64 && \
chmod +x skaffold && sudo mv skaffold /usr/local/bin
```

If you don't have homebrew or MacPorts, check out the following links for install instructions:

    * Minikube: https://minikube.sigs.k8s.io/docs/start/
    * Skaffold: https://skaffold.dev/docs/install/


### On Linux:
```
     curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-linux-amd64 && \
     chmod +x skaffold && sudo mv skaffold /usr/local/bin

    curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl \
    && sudo install kubectl /usr/local/bin && rm kubectl

    curl https://baltocdn.com/helm/signing.asc | sudo apt-key add - && \
    sudo apt-get install apt-transport-https --yes && \
    echo "deb https://baltocdn.com/helm/stable/debian/ all main" | sudo tee /etc/apt/sources.list.d/helm-stable-debian.list && \
    sudo apt-get update && \
    sudo apt-get install helm

    curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash

    curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add - && \
    sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu focal stable" && \
    sudo apt update && \
    sudo apt install docker-ce && \
    sudo usermod -aG docker ${USER}

    curl -LO https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64 \
    && sudo install minikube-linux-amd64 /usr/local/bin/minikube && rm minikube-linux-amd64
```

## Running Minikube locally

minikube can easily be run just by typing:

```
minikube start
```

While we've tested on minikube with default options, and we get a working demo environment, we recommend allocating 3-4 cpus and 8192MB of RAM. This repo has been tested both with the hyperkit driver:

```
minikube start --cpus=4 --memory=8192mb --vm-driver=hyperkit
```

as well as with the docker driver:

```
minikube start --cpus=4 --memory=8192mb --vm-driver=docker
```

minikube and skaffold don't always get along too well when it comes to image caching, so it's important to ensure that 
skaffold uses the docker environment within minikube. Luckily, this is pretty easy to do. Simple run:

```
eval $(minikube docker-env)
```

this will set a few environment variables that will instruct skaffold to use the minikube docker env, rather than the
systems docker env. 

## Running a Demo Environment


**_NOTE_**: This has changed - with the move to using configmaps instead of custom images, an additional setup has been added.

### Setup.sh 
The new step is the addition of a script called `setup.sh`, which sets up all of the ConfigMap and Secret objects that need to be in place. by default, it will set the objects up in the default namespace. The script supports the following options:
* `-n <namespace>` - kubernetes namespace to install in. This defaults to `default`
* `-p <password>` - admin password to set for the LogStream, Grafana, Splunk and InfluxDB instances. This defaults to `cribldemo`.
* `-b <branch>` - to set the branch for the ls-deploy job to run from. This only really necessary when testing changes in an alternate branch. This defaults to `master`
* `-a <authsecret>` - will update the default service account on the specified (or `default` if not specified) namespace to use the specified authsecret. This has no default.
* `-s` - enable AppScope in the resulting demo environment - AppScope needs to be installed on the cluster first (see [APPSCOPE.md](APPSCOPE.md) for details). 

**_NOTE_**: Dockerhub, as of November 2020, has implemented fairly restrictive rate limiting for anonymous access. If you see errors that refer to too many requests, you'll need to sign up for a free docker account, create an authsecret (see the kubernetes.io [doc](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account) for details on this), and use the -a option listed above to eliminate those errors. 

### Steps to Run

By default, the repo is set up to run on a fairly robust kubernetes cluster. To be able to run on a typical laptop, we needed to reduce the cpu/memory requirements. To do that, a combination of skaffold's profiles and kustomize are used to create options for a smaller footprint. 

The `-p dev` argument to skaffold invokes the "dev" profile, which minimizes the footprint of the deployment. If that's omitted, the services for cribl, splunk and grafana will all attempt to create load balancers and require significantly more horsepower from the hosting machine.


To run the demo LOCALLY on minikube (again, on a Mac):
```
    minikube start --cpus=4 --memory=8192mb  # Starts Minikube
    ./scripts/setup.sh                       # Sets up Configmaps
    skaffold dev --port-forward=true -p dev  # Starts the demo environment
```

Now, you can access Cribl at http://localhost:9000 with username `admin` password `cribldemo` (or the password you specified with the setup -p option). 

### Oddities in Execution


## Accessing Services in the Demo environment

Using the port-forwarding capability in skaffold (as seen above), will yield 4 services that you can access:

|Service|URL|Purpose|
|-------|---|-------|
|cribl|[http://localhost:9000](http://localhost:9000)|The Cribl LogStream master UI|
|splunk|[http://localhost:8000](http://localhost:8000)|The Splunk UI|
|influxdb2|[http://localhost:8086](http://localhost:8086)|The InfluxDB UI|
|grafana|[http://localhost:4000](http://localhost:4000)|The Grafana UI|


## Profiles<a name=profiles></a>

We have alternate profiles in the skaffold.yaml file:

* dev - this reduced the memory load of the environment by eliminating the resource allocations for each pod. The user facing services (logstream master, splunk, grafan and influxdb2) all will run in NodePort mode. It will also remove the data creation jobs (gogen-datacollection-syslog and cribl-sa), and the AWS worker group (aka cribl-w2)
* minimal - this also reduces the memory load through eliminated resource allocations and removes the aws worker group (aka cribl-w2). It also removes influxdb, grafana and splunk. This can be useful when looking to prototype something only in logstream. Beware when running in this profile, since incoming data will cause back pressure on the sources.
* nogen - this also reduces the memory load through eliminated resource allocations and removes the aws worker group (aka cribl-w2), and removes all of the gogen data generators from the deployed environment. 
* forward - this profile can be combined with any of the other profiles to forward the user facing services to 0.0.0.0 (instead of the default 127.0.0.1), allowing for access to it from other systems - useful when you're developing on an AWS EC2 instance with no GUI. 


## EKS Deployment

At Cribl, we run our standard demo environments on an AWS EKS cluster. If you would like to deploy this on EKS, see the [EKS-DEPLOY.md](EKS-DEPLOY.md) file. 

## Contributing to the cribl-demo project

If you want to contribute to the repo, see the [CONTRIBUTING.md](CONTRIBUTING.md) file.

# Reference Material

* [Minikube Tutorial](https://kubernetes.io/docs/tutorials/hello-minikube/) on kubernetes.io
* [Minikube Github Project](https://github.com/kubernetes/minikube)


CHANGES.md
# Changes to the Demo Environment

## June 4, 2021

### Demo Changes
* The three apache log based pipelines in dc1-logs (enrich, logs_to_metrics, and sample_and_filter) are now part of the "Apache Logs" Pack. To show those pipelines, go to the routing table, and find the "Apache Access Logs" route:
![UI Screenshot](img/apache_pack.png)
You'll see that In the pipeline/output section, there is now a "pack" icon next to the pack name. You can click on the pack name link, which takes you to the pack interface.
* The firewall_geoip_enrich pipeline (dc1-logs group) has now also moved to a pack, named pan-firewall-traffic.
* The vpcflow_geoip_enrich pipeline (aws group) has also moved to a pack, named vpcflow-parse.
* The demo is now Appscope ready - it is not enabled by default, but can be - see the [APPSCOPE.md](APPSCOPE.md) document. 
* The trim_json pipeline and route has been moved back to the dc1-logs group.
* AppScope "skeleton" is setup for all services but influx (disabled due to TLS bug with AppScope 0.6.1, will renable after 0.7.0). AppScope is not enabled by default (see [APPSCOPE.md](APPSCOPE.md) for details on enabling it). It is enabled on the Cribl official demo environments.
* New `AppScope Metrics` dashboard in Influx shows stats for the environment by service. This will evolve over time. 
* `Windows_Events.crbl` pack deployed to dc1-logs - currently no data generation for it (demoable by sample file).

### Structural Changes
* One of the big problem areas in running the demo setup has been the disparity between helm deployment and manifest deployment, specifically with the setup of influxdb - job timeouts have been the biggest contributor to standup failure. In this version, instead of maintaining a copy of helm charts and using the helm capability to install them, we're making use of a pattern where we "hydrate" the chart into a manifest regularly and commit that, which has minimize the use of helm in the setup. 
* Additionally, far fewer custom images are being built as a part of standup. We've made use of the new features in the cribl helm charts that allow configmaps to be referenced. Unfortunately, this necessitates a "setup" script be run prior to skaffold being called, but the benefits of this approach outweigh the negatives. 
* The data collection data generation (gogen-datacollection-syslog and cribl-sa) have now been converted to K8s jobs.
* The cli-deploy and api-deploy scripts have been deprecated in favor of a new ls-deploy-job K8s job that not only commits and deploys initial configuration, but also handles installation of Packs, a new feature in LogStream 3.0.
* Appscope is now an option - you can now enable appscope for the namespace you're running in, and have it deliver it's data to the cribl-w2 instance. This is *not* available on the `dev`, `minimal` or `nogen` profiles, as it requires the entire pod set in place.
* Profiles have changed a little bit, and can be combined if necessary - see the [README.md](README.md#profiles) for descriptions and usage details.

	

  

## Feb 5, 2021

### Structural Changes

* Moved cribl-demo to only run on Kubernetes, retiring the docker-compose framework, as well as the env-builder based mastering of the repo. 
* Worker Group Changes - To show that worker group definition is somewhat arbitrary, based on the needs of the customer, and not necessarily based on any functional best practice.

	* Renamed worker groups:

      |Old|New|
      |---|---|
      |logs|dc1-logs|
      |metrics|dc1-metrics|

    * Introduced new worker group: aws, and moved the Trim Big JSON demo flow to it.
    
* Added vpcflow logs data, collector and pipeline to the aws worker group. This also includes a redis lookup of aws account id to name, and the pipeline is modeled after the firewall_geoip_enrich pipeline in dc1-logs, implying that you can use logstream to unify two disparate types of log data (PAN firewall + VPC flowlogs)

* Replaced old demo redis_lookup function in enrich pipeline with the new (2.4+) redis function.

* Created a new branch `docker-legacy` to support people transitioning from the docker-compose based flow to the kubernetes flow, which will be present until 4/1/2021 as a "transition period"

* Cleanup and bug fixes as requested by the SE team and others.

# Topology
![Cribl-Demo Topology](img/cribl-demo-topology.png)


deploy-eks.py
#!/usr/bin/env python3

from optparse import OptionParser
from kubernetes import client, config
from yaml import load, dump
import sys
import boto3
import json
import botocore.exceptions
import subprocess
import os
import re
import time
import docker
import base64
import urllib3
import collections
import json
import tempfile
import shutil
import tarfile
from urllib.parse import quote_plus



try:
    from yaml import CLoader as Loader, CDumper as Dumper
except ImportError:
    from yaml import Loader, Dumper

#config objects
services = { 
  "cribl": "Cribl LogStream UI", 
  "grafana": "Grafana Visualization Tool",  
  "influxdb2": "InfluxData's InfluxDB v2",
  "splunk": "Splunk UI"}
allowed_ports = [ 3000, 8000, 8086, 9000 ]

def gen_demo_config(params,options):
  configmap = "demo-config"
  print("Setting up Auth ConfigMap")
  # Delete first, in case there's already one present.
    # Check for and delete if already present.
  kbctlcheck = "kubectl get configmap %s -n %s >/dev/null 2>&1" % (configmap, options.ns)
  kbctlchkout = subprocess.call(kbctlcheck, shell=True)
  if (kbctlchkout == 0):
    print("Deleting Existing Auth Configmap")
    kbctldel = "kubectl delete configmap %s -n %s" % (configmap, options.ns)
    kubectlout = subprocess.call(kbctldel,  shell=True)

  #f = tempfile.TemporaryDirectory(dir="/var/tmp")
  dir = tempfile.mkdtemp(dir = "/var/tmp")

  print("Temp Dir: %s" % dir)
  # Write the users.json file
  with open("%s/users.json" % dir, "w") as fp:

    for cred in params['creds']:
      thiscred = params['creds'][cred];
      if cred == "admin":
        fp.write ("{\"username\":\"admin\",\"first\":\"admin\",\"last\":\"admin\",\"email\":\"admin\",\"roles\":[\"admin\"],\"password\":\"%s\"}\n" % thiscred['password'])
      else:
        fp.write("{\"first\":\"%s\",\"last\":\"%s\",\"email\":\"%s\",\"roles\":[%s],\"username\":\"%s\",\"password\":\"%s\"}\n" % (thiscred['first'], thiscred['last'],thiscred['email'],re.sub(r',', '","', re.sub(r'^|$','"',thiscred['rolestring'])),cred,thiscred['password']))
  # Write the branch file
  with open("%s/branch" % dir, "w") as fp:
    if "branch" in params:
      fp.write(params['branch'])
    else:
      fp.write("master")
  # Write the admin.pass file
  with open("%s/admin.pass" % dir, "w") as fp:
    if "creds" in params and "admin" in params['creds'] and "password" in params['creds']['admin']:
      fp.write(params['creds']['admin']['password'])
    else:
      fp.write("cribl-demo")
  
  # Create the configmap (need to add error handling)
  kbctlcr = "kubectl create configmap demo-config -n %s --from-file=%s " % (options.ns, dir)
  kubectlout = subprocess.call(kbctlcr,  shell=True)
  if (kubectlout != 0):
    print("Auth Configmap Failed: %s" % kubectlout)
  
  print("Creating general secret")
  secret = "demo-admin"
  
  # Check for and delete if already present.
  kbctlcheck = "kubectl get secret %s -n %s " % (secret, options.ns)
  kbctlchkout = subprocess.call(kbctlcheck, shell=True)
  if (kbctlchkout == 0):
    kbctldel = "kubectl delete secret %s -n %s" % (secret, options.ns)
    kubectlout = subprocess.call(kbctldel,  shell=True)
  kbctlsecret = "kubectl create secret generic %s -n %s --from-literal=adminpass=%s" % (secret, options.ns, params['creds']['admin']['password'])
  kbctlout = subprocess.call(kbctlsecret,  shell=True)
  if (kbctlout != 0):
    print("Error creating Secret")
  shutil.rmtree(dir)

def run_setup(params,options):
  cmd="./scripts/setup.sh"

  if not os.path.exists(cmd):
    print("Error: Not in the Repo Top Level")
    sys.exit(255)

  #CRIBL_TAG=next ./scripts/setup.sh -a dhub -n logstreamnext
  if options.pullsecret:
    cmd += " -a %s" % options.pullsecret

  #cmd += " -n %s -s" % options.ns
  cmd += " -n %s " % options.ns
  print ("Running: %s" % cmd)
  cmdout = subprocess.call(cmd,  shell=True)
  print ("Setup Run Return: %s" % cmdout)
  

  
   


def gen_master_config(params,options):
  configmap = "master-config"
  # This configmap includes the files in cribl/master/local/cribl
  # direct call to kubectl to create the configmap from those files.
  dir="cribl/master/local/cribl/"

  if not os.path.exists(dir):
    print("Error: Not in the Repo Top Level")
    sys.exit(255)

  print("Setting up ConfigMap %s" % configmap)
  # Check for and delete if already present.
  kbctlcheck = "kubectl get configmap %s -n %s > /dev/null 2>&1" % (configmap, options.ns)
  kbctlchkout = subprocess.call(kbctlcheck, shell=True)
  if (kbctlchkout == 0):
    kbctldel = "kubectl delete configmap %s -n %s" % (configmap, options.ns)
    kubectlout = subprocess.call(kbctldel,  shell=True)

  # Create the configmap (need to add error handling)
  kbctlcr = "kubectl create configmap %s -n %s --from-file=%s" % (configmap, options.ns, dir)
  kubectlout = subprocess.call(kbctlcr,  shell=True)
  if (kubectlout != 0):
    print("Failed to Create ConfigMap %s: %s" % (configmap, kubectlout))

def gen_group_config(params,options):
  configmap = "group-config"
  running_dir = os.getcwd()
  conf_dir="cribl/master/groups"

  if not os.path.exists(conf_dir):
    print("Error: Not in the Repo Top Level")
    sys.exit(255)
  #change to the source directory
  os.chdir(conf_dir)
  # Check for and delete if already present.
  kbctlcheck = "kubectl get configmap %s -n %s > /dev/null 2>&1" % (configmap, options.ns)
  kbctlchkout = subprocess.call(kbctlcheck, shell=True)
  if (kbctlchkout == 0):
    kbctldel = "kubectl delete configmap %s -n %s" % (configmap, options.ns)
    kubectlout = subprocess.call(kbctldel,  shell=True)

  source_contents = os.listdir(".")
  dir = tempfile.mkdtemp(dir = "/var/tmp")
  print("Temp Dir: %s" % dir)

  for source_dir in source_contents:
    if os.path.isdir(source_dir):
  
      print("Tarring up %s" % source_dir)
      with tarfile.open("%s/groups-%s.tgz" % (dir,source_dir), "w:gz") as tar:
        tar.add(source_dir)

  # Create the configmap (need to add error handling)
  kbctlcr = "kubectl create configmap %s -n %s --from-file=%s" % (configmap, options.ns, dir)
  kubectlout = subprocess.call(kbctlcr,  shell=True)
  if (kubectlout != 0):
    print("Failed to Create ConfigMap %s: %s" % (configmap, kubectlout))

  # Change back to the original directory
  os.chdir(running_dir)
  shutil.rmtree(dir)
  
def gen_sa_config(params,options):
  configmap = "sa-config"
  running_dir = os.getcwd()
  conf_dir="cribl/sa"

  if not os.path.exists(conf_dir):
    print("Error: Not in the Repo Top Level")
    sys.exit(255)
  #change to the source directory
  os.chdir(conf_dir)
  # Check for and delete if already present.
  kbctlcheck = "kubectl get configmap %s -n %s > /dev/null 2>&1" % (configmap, options.ns)
  kbctlchkout = subprocess.call(kbctlcheck, shell=True)
  if (kbctlchkout == 0):
    kbctldel = "kubectl delete configmap %s -n %s" % (configmap, options.ns)
    kubectlout = subprocess.call(kbctldel,  shell=True)

  source_contents = ['cribl']
  dir = tempfile.mkdtemp(dir = "/var/tmp")
  print("Temp Dir: %s" % dir)

  for source_dir in source_contents:
    if os.path.isdir(source_dir):
  
      print("Tarring up %s" % source_dir)
      with tarfile.open("%s/sa.tgz" % dir, "w:gz") as tar:
        tar.add(source_dir)

  # Create the configmap (need to add error handling)
  kbctlcr = "kubectl create configmap %s -n %s --from-file=%s" % (configmap, options.ns, dir)
  kubectlout = subprocess.call(kbctlcr,  shell=True)
  if (kubectlout != 0):
    print("Failed to Create ConfigMap %s: %s" % (configmap, kubectlout))

  shutil.rmtree(dir)
  # Change back to the original directory
  os.chdir(running_dir)


def dict_merge(dct, merge_dct):
    """ Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param dct: dict onto which the merge is executed
    :param merge_dct: dct merged into dct
    :return: None
    """
    for k, v in merge_dct.items():
        if (k in dct and isinstance(dct[k], dict)
                and isinstance(merge_dct[k], collections.abc.Mapping)):
            dict_merge(dct[k], merge_dct[k])
        else:
            dct[k] = merge_dct[k]


def param_parse(paramtmp, path):
  tmpparam = {}

  # Write a simpler dict to work with...
  for i in paramtmp:
    #print("I: %s" % i)
    i['Name'] = i['Name'].replace(path + "/","")

    if "/" in i['Name']:
      tmp = i['Name'].split("/")
      #print("Break")
      myparam = tmpparam
      for ent in range(0, len(tmp)-1):

        #print("Entry: %s" % tmp[ent])
        if tmp[ent] not in myparam:
          myparam[tmp[ent]] = {}
        myparam = myparam[tmp[ent]]
      myparam[tmp[len(tmp)-1]]=i['Value']

    else:
      #print("Param: %s - %s" % (i['Name'], i['Value']))
      tmpparam[i['Name']] = i['Value']

  return(tmpparam)


def pull_params(parampath, namespace):
  ssm = boto3.client("ssm")
  params = {}

  fullpath = "%s/%s" % (parampath, namespace)
  print("Full: %s" % fullpath)

  try:
    pass_param = ssm.get_parameters_by_path(Path=fullpath, Recursive=True, WithDecryption=True, MaxResults=10)
  except botocore.exceptions.ClientError as e:
    if e.response['Error']['Code'] == 'ParameterNotFound':
      print('Parameter tree does not exist, Will use command line args exclusively')
    print(e)

  #params.update(param_parse(pass_param['Parameters']))
  dict_merge(params, param_parse(pass_param['Parameters'], fullpath))

  while("NextToken" in pass_param):
    try:
      pass_param = ssm.get_parameters_by_path(Path=fullpath, Recursive=True, WithDecryption=True, MaxResults=10, NextToken=pass_param['NextToken'])
    except botocore.exceptions.ClientError as e:
      if e.response['Error']['Code'] == 'ParameterNotFound':
        print('Parameter tree does not exist, Will use command line args exclusively')
      print(e)

    dict_merge(params, param_parse(pass_param['Parameters'], fullpath))
  return params

def base64_encode(string):
    """
    Removes any `=` used as padding from the encoded string.
    """
    encoded = base64.urlsafe_b64encode(string)
    return encoded.rstrip(b"=")


def base64_decode(string):
    """
    Adds back in the required padding before decoding.
    """
    padding = 4 - (len(string) % 4)
    string = string + ("=" * padding)
    return base64.urlsafe_b64decode(string)

def get_cluster_name():
  kubectlout = subprocess.run(["kubectl", "config", "current-context"],capture_output=True)
  cluster = kubectlout.stdout.decode()[:-1]
  return cluster

def docker_login(options, acct):
  client = docker.from_env()
  ecr = boto3.client('ecr')
  token = ecr.get_authorization_token()

  username, password = base64.b64decode(token['authorizationData'][0]['authorizationToken']).decode().split(':')
  registry = token['authorizationData'][0]['proxyEndpoint']
  #print("Token: %s - %s - %s" % (username,password,registry))
  resp = client.login(username, password, registry=registry, reauth=True)
  logincmd = "aws ecr get-login-password | docker login --username AWS --password-stdin %s" % registry
  subprocess.run(["docker", "login", "--username", "AWS", "--password-stdin", registry], input=password.encode('utf-8'))
  print(resp['Status'])


# Functions
# Set up ECR Repos for all the images in the skaffold.yaml file.
def setup_ecr(options,acct):
  ecr = boto3.client('ecr') 

  print("Ensuring ECR Repo is Setup...")
  with open('skaffold.yaml') as skf:
    skafcfg = load(skf, Loader=Loader)
    #print (json.dumps(skafcfg['build']['artifacts']))
    for art in skafcfg['build']['artifacts']:

      try:
        ecr.create_repository(repositoryName="%s/%s" % (options.repohead, art['image']))
      except botocore.exceptions.ClientError as error:
        if error.response['Error']['Code'] == "RepositoryAlreadyExistsException":
          #print("Repo Exists")
          pass
        else:
          print("Unhandled Error: %s" % error.response['Error']['Code'])

  print("Done")
  docker_login(options,acct)

# Check that the specified namespace exists, and create it if it doesn't
def check_namespace(options, kubeclient):
  nslist = kubeclient.list_namespace()
  found = 0
  for ns in nslist.items:
    if (ns.metadata.name == options.ns):
      found = 1
      print (ns.metadata.name)
  
  if found == 0:
    print("Need to create the namespace")
    kubeclient.create_namespace(client.V1Namespace(metadata=client.V1ObjectMeta(name=options.ns)))
  
  patch = {'metadata': { 
          'annotations': {'description': options.description}}}
  kubeclient.patch_namespace(options.ns, patch)

# Lookup the zoneid of the hosted zone - die if it doesn't exist.
def get_hosted_zone(options):
  r53 = boto3.client("route53")

  print ("Finding Domain %s" % options.domain)
  retzones = r53.list_hosted_zones_by_name(DNSName=options.domain)
  for zone in retzones['HostedZones']:
    print ("Name; %s" % zone['Name'])
    if (zone['Name'] == "%s." % options.domain):
      return zone['Id']  
  
    print("No Zone Found - Not Continuing")
    sys.exit(0)

# Set up Command Line Parsing
parser = OptionParser()
parser.add_option("-n", "--namespace", dest="ns", default="default", help="Namespace to Interrogate")
parser.add_option("-d", "--domain", dest="domain", default="demo.cribl.io", help="Hosted Zone to Use")
parser.add_option("-b", "--pull-secret", dest="pullsecret", help="Pull Secret to use for the namespace default service account")
parser.add_option("-r", "--region", dest="region", default="us-west-2", help="AWS Region to deploy to")
parser.add_option("-s", "--ssm-path", dest="ssmpath", default="/cribl/demo", help="SSM path for Environment options")
parser.add_option("-a", "--description", dest="description", default="Demo Environment")
parser.add_option("-c", "--container-repo-head", dest="repohead", default="cribl-demo", help="ECR Repo top level")
parser.add_option("-p", "--profile", dest="profile", help="Skaffold Profile to run with")
(options, args) = parser.parse_args()


# Set default env variables
if "CRIBL_TAG" not in os.environ:
  os.environ['CRIBL_TAG'] = "latest"

# Set up AWS objects
s3 = boto3.client('s3')
sts = boto3.client('sts')
r53 = boto3.client("route53")
ssm = boto3.client("ssm")

parampath=options.ssmpath + "/" + options.ns
#print("Parampath: %s" % parampath)
parameters = pull_params(options.ssmpath, options.ns)

chpass=False


if "creds" in parameters and "admin" in parameters["creds"] and "password" in parameters["creds"]["admin"]:
  chpass = True
  print('Credential param exists, using it')
else: 
  chpass = False
  print('Using Default Credentials')

if ("repo" in parameters):
  print("Found Repo")
  options.repohead = parameters['repo']

if ("description" in parameters):
  options.description = parameters['description']

if ("domain" in parameters):
  options.domain = parameters['domain']

if ("profile" in parameters):
  options.profile = parameters['profile']

if ("tag" in parameters):
  os.environ['CRIBL_TAG'] = parameters['tag']

#print("Options: %s" % options)

if chpass:
  cmd="perl -pi.bak -e 's{cribldemo([\"\\\]+)}{" + parameters['creds']['admin']['password'] + "$1}g;' ./grafana/grafana.k8s.yml"
  rval = subprocess.call(cmd,  shell=True)
  if rval == 0:
    print("Password Set Succeeded")

# print("before call: %s" % parameters)
run_setup(parameters,options)
gen_demo_config(parameters,options)
#gen_master_config(parameters,options)
#gen_group_config(parameters,options)
#gen_sa_config(parameters,options)

# get acct id and hosted zone id
acct = sts.get_caller_identity()
(id,email) = acct['UserId'].split(':')
mtch = re.match(r'@', email)
if mtch:
  options.description += "<br><font size=-1>(Created by <a href=mailto:%s>%s</a>)</font>" % (email, email)
else:
  options.description += "<br><font size=-1>(Created by Automation)</font>"
zoneid = get_hosted_zone(options)

# Make sure the ECR repos are setup.
setup_ecr(options,acct)

# Setup the K8s API Client
config.load_kube_config()
kubeclient = client.CoreV1Api()

# ensure that the specified namespace is set up
check_namespace(options, kubeclient)

# Set up the SKAFFOLD env var
os.environ['SKAFFOLD_DEFAULT_REPO'] = "%s.dkr.ecr.%s.amazonaws.com/%s" % (acct['Account'], options.region, options.repohead)

# Run skaffold build with the namespace as the tag (this allows us to have different images for different deployments/namespaces)
skaffbuildcall = "skaffold build --tag=%s" % options.ns
skaffdeploycall = "skaffold deploy --status-check --tag=%s -n %s" % (options.ns, options.ns)
if (options.profile):
  skaffbuildcall = "skaffold build --tag=%s --profile=%s" % (options.ns, options.profile)
  skaffdeploycall = "skaffold deploy --status-check --tag=%s --profile=%s -n %s" % (options.ns, options.profile, options.ns)

rval = subprocess.call(skaffbuildcall,  shell=True)
if rval == 0:
  print("Skaffold Build Succeeded")
else:
  print("Skaffold Build Failed")
  sys.exit(rval)

# Run Skaffold Deploy with the namespace as tag and namespace to deploy in
rval = subprocess.call(skaffdeploycall,  shell=True)
if rval == 0:
  print("Skaffold Deploy Succeeded")
else:
  print("Skaffold Deploy Failed")
  sys.exit(rval)

time.sleep(60)

# Setup the R53 JSON
chgbatch = { "Comment": "Upserting for the %s.%s domain" % (options.ns, options.domain), "Changes": []}



# Reverse the specified domain to make the bucket name
domrev = options.domain.split('.')
domrev.reverse()
revhost = ".".join(domrev)

print("Creating HTML Service View")

style = '''
<style>
      img {
        max-width: 15%;
        height: auto;
      }
      body {
        background-color: black;
        color: white;
      }
      h1, h2, h3, h4, h5 {
        color: white;
      }
      table {
        width:90%;
      }
      table, th, td {
        border: 1px solid #C9D0D4;
        text-align: left;
        padding: 10px;
      }
    </style>
'''
# Get the Service info, and generate an index.html to post for the namespace.
htmlout = """
<HTML>
  <HEAD>
    <TITLE>Services in %s</TITLE>
    <link rel=stylesheet href=https://cribl.io/wp-content/themes/cribl/assets/css/main.css?ver=1608242659>
    %s
  
  </HEAD>
  <BODY>
    <H2><img height=100 src=https://cribl.io/wp-content/themes/cribl/assets/images/logo-cribl-new.svg><br>Services in %s</H2>
    <p>%s</p>
    <table>
    <tr><th>Service</th><th>Service Description</th></tr>
""" % (options.ns, style, options.ns, options.description)

for svc in services.keys():
  print ("Checking svc %s service" % svc)
  ret = kubeclient.read_namespaced_service(svc, options.ns)
  for host in ret.status.load_balancer.ingress:
    chg = { "Action": "UPSERT", "ResourceRecordSet": { "Name": "%s.%s.%s" % ( svc, options.ns, options.domain ), "Type": "CNAME", "TTL": 300, "ResourceRecords": [ { "Value": host.hostname } ] } }
    chgbatch['Changes'].append(chg)
    for port in ret.spec.ports:
      if port.port in allowed_ports:
        htmlout += "<tr><td><A HREF=http://%s.%s.%s:%d/>%s</A></td><td>%s</td></tr>" % ( svc, options.ns, options.domain, port.port, svc, services[svc])

htmlout += "</table></body></HTML>"

htmlb = htmlout.encode("utf8")

desc = options.description
#desc=options.description.replace("@","-").replace("<br>"," - ")
#desc = re.sub(r'Created by A[^:]+:', 'Created by', desc)
#tagset=quote_plus({"namespace-description":  options.description})
#options.description+="na"h
tdat = base64.urlsafe_b64encode(options.description.encode('utf-8')).decode('utf-8')
cluster_name = get_cluster_name()
cname = base64.urlsafe_b64encode(cluster_name.encode('utf-8')).decode('utf-8')
tagset="cluster=%s&namespace-description=%s" % (cname, tdat)

#print("Tagset: %s" % tagset)

# put the index.html file up.
resp = s3.put_object(Bucket=revhost, Body=htmlb, Key="ns-%s/index.html" % options.ns, ACL='public-read', ContentType='text/html', Tagging=tagset)
print("Done")

print ("Updating R53")
response = r53.change_resource_record_sets(HostedZoneId=zoneid, ChangeBatch = chgbatch)
print("%s - %s" % (response['ChangeInfo']['Status'], response['ChangeInfo']['Comment']))

if (chpass):
  cmd="git checkout ./grafana/grafana.k8s.yml"
  rval = subprocess.call(cmd,  shell=True)
  if rval == 0:
    print("Password Unset Succeeded")
